File: C11/백준/Silver/1991. 트리 순회/트리 순회.c
#include <stdio.h>

typedef struct {
    char value;
    char left;
    char right;
} Node;

Node tree[26];
int N;

void preorder(char node) {
    if (node == '.') return;
    printf("%c", node);
    preorder(tree[node - 'A'].left);
    preorder(tree[node - 'A'].right);
}

void inorder(char node) {
    if (node == '.') return;
    inorder(tree[node - 'A'].left);
    printf("%c", node);
    inorder(tree[node - 'A'].right);
}

void postorder(char node) {
    if (node == '.') return;
    postorder(tree[node - 'A'].left);
    postorder(tree[node - 'A'].right);
    printf("%c", node);
}

int main() {
    scanf("%d", &N);
    for (int i = 0; i < N; i++) {
        char node, left, right;
        scanf(" %c %c %c", &node, &left, &right);
        tree[node - 'A'].value = node;
        tree[node - 'A'].left = left;
        tree[node - 'A'].right = right;
    }
    preorder('A');
    printf("\n");
    inorder('A');
    printf("\n");
    postorder('A');
    printf("\n");
    return 0;
}

File: Java/백준/Silver/1991. 트리 순회/트리 순회.java
import java.io.*;
import java.util.*;

public class Main {
    static int N;
    static Map<Character, Node> tree = new HashMap<>();

    static class Node {
        char value;
        char left;
        char right;
        Node(char value, char left, char right) {
            this.value = value;
            this.left = left;
            this.right = right;
        }
    }

    static StringBuilder pre = new StringBuilder();
    static StringBuilder in = new StringBuilder();
    static StringBuilder post = new StringBuilder();

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        N = Integer.parseInt(br.readLine());
        for (int i = 0; i < N; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            char node = st.nextToken().charAt(0);
            char left = st.nextToken().charAt(0);
            char right = st.nextToken().charAt(0);
            tree.put(node, new Node(node, left, right));
        }
        preorder('A');
        inorder('A');
        postorder('A');
        System.out.println(pre.toString());
        System.out.println(in.toString());
        System.out.println(post.toString());
    }

    static void preorder(char node) {
        if (node == '.') return;
        pre.append(node);
        preorder(tree.get(node).left);
        preorder(tree.get(node).right);
    }

    static void inorder(char node) {
        if (node == '.') return;
        inorder(tree.get(node).left);
        in.append(node);
        inorder(tree.get(node).right);
    }

    static void postorder(char node) {
        if (node == '.') return;
        postorder(tree.get(node).left);
        postorder(tree.get(node).right);
        post.append(node);
    }
}